minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,796 @@
+        return self.node_kind is NodeKind.HORIZONTAL
+
+    def neighbor_kind(
+        self,
+        other: ZNode,
+    ) -> EdgeKind | None:
+        """Returns the kind of edge between self and other;
+        EdgeKind if there is an edge in perfect Zephyr between them or None.
+
+        Args:
+            other (ZNode): a ZNode
+
+        Returns:
+            EdgeKind | None: The edge kind between self and other in perfect Zephyr or None if there is no edge between in perfect Zephyr.
+        """
+        if not isinstance(other, ZNode):
+            other = ZNode(other)
+        if self._shape != other._shape:
+            return None
+        coord1 = self._ccoord
+        coord2 = other._ccoord
+        x1, y1 = coord1.x, coord1.y
+        x2, y2 = coord2.x, coord2.y
+        if abs(x1 - x2) == abs(y1 - y2) == 1:
+            return EdgeKind.INTERNAL
+        if x1 % 2 != x2 % 2:
+            return None
+        if coord1.k != coord2.k:  # odd, external neighbors only on the same k
+            return None
+        if self.node_kind is NodeKind.VERTICAL:  # self vertical
+            if x1 != x2:  # odd, external neighbors only on the same vertical lines
+                return None
+            diff_y = abs(y1 - y2)
+            return EdgeKind.ODD if diff_y == 2 else EdgeKind.EXTERNAL if diff_y == 4 else None
+
+        # else, self is horizontal
+        if y1 != y2:  # odd, external neighbors only on the same horizontal lines
+            return None
+        diff_x = abs(x1 - x2)
+        return EdgeKind.ODD if diff_x == 2 else EdgeKind.EXTERNAL if diff_x == 4 else None
+
+    def internal_neighbors_generator(
+        self,
+        where: Callable[[CartesianCoord | ZephyrCoord], bool] = lambda coord: True,
+    ) -> Generator[ZNode]:
+        """Generator of internal neighbors of self when restricted by `where`.
+
+        Args:
+            where (Callable[[CartesianCoord | ZephyrCoord], bool], optional):
+                A coordinate filter. Applies to `ccoord` if `self.convert_to_z` is False,
+                or to `zcoord` if `self.convert_to_z` is True. Defaults to `lambda coord: True`.
+
+        Yields:
+            ZNode: Internal neighbors of self when restricted by `where`.
+        """
+        x, y, _ = self._ccoord
+        convert = self.convert_to_z
+        k_vals = [None] if self._shape.t is None else range(self._shape.t)
+        for i, j, k in product((-1, 1), (-1, 1), k_vals):
+            ccoord = CartesianCoord(x=x + i, y=y + j, k=k)
+            coord = ccoord if not convert else cartesian_to_zephyr(ccoord)
+            if not where(coord):
+                continue
+            try:
+                yield ZNode(
+                    coord=ccoord,
+                    shape=self._shape,
+                    convert_to_z=convert,
+                )
+            except GeneratorExit:
+                raise
+            except (TypeError, ValueError):
+                pass
```

Comment:
I don't think you necessarily need this try-except block. You shouldn't need to catch a `GeneratorExit` in this case, and the type error should be handled earlier (i.e., `ccoord` and `self._shape` should already have the correct type). The value check, if needed, I'd have earlier instead of wrapping it into a try-except.
```suggestion

            yield ZNode(
                coord=ccoord,
                shape=self._shape,
                convert_to_z=convert,
            )
```

Reply:
Agree that `TypeError` is unnecessary. However, the value check is needed, and if we want to do it earlier, it requires all the checks in `ccoord.setter` of `ZNode` to be repeated here.

Reply:
Done in a56aa8453ea0915c9c4b824e5606939d762a3561